### PR TITLE
Cascade classifier

### PR DESCRIPTION
**Pull request by bitonic**

### CODE REVIEW

File: test/test.hs [hunk 1]
```diff
@@ -291,6 +295,28 @@ testHoughLinesP = do
...
+testNewCascadeClassifierAlgorithm :: HU.Assertion
+testNewCascadeClassifierAlgorithm = do
+  mbCC <- newCascadeClassifier "/this/is/bogus.xml"
+  case mbCC of
+    Nothing -> return ()
+    Just _cc -> fail "expected Nothing from newCascadeClassifier"
+
+testCascadeClassifierDetectMultiScaleArnold :: HU.Assertion
+testCascadeClassifierDetectMultiScaleArnold = do
+  Just ccFrontal <- newCascadeClassifier "/usr/local/share/OpenCV/haarcascades/haarcascade_frontalface_default.xml"
```
Comment (bitonic) [inline test/test.hs]: This is hardcoded as it is on my machine, but we should change it to some nix location, if possible.

File: src/OpenCV/ImgProc/CascadeClassifier.hs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                [CU.exp| void { delete $(CascadeClassifier * ptr) }|]
+
+-- | Create a new cascade classifier. Returns 'Nothing' if the classifier
+-- is empty after initialization. This usually means that the file could
+-- not be loaded (e.g. it doesn't exist, is corrupt, etc.)
+newCascadeClassifier :: FilePath -> IO (Maybe CascadeClassifier)
+newCascadeClassifier fp = do
+  cc <- withCString fp $ \c'fp -> fromPtr
+    [CU.exp| CascadeClassifier * { new CascadeClassifier(cv::String($(const char * c'fp))) } |]
+  -- TODO: empty() seems to return bogus numbers when the classifier is not
```
Comment (bitonic) [inline src/OpenCV/ImgProc/CascadeClassifier.hs]: I'm not sure why this happens, but I don't think it is my fault.

**Comment by bitonic** [comment]
It would be nice to add an example to be shown in the doc, for example just putting the rectangles on Arnold's face, showing the detected eyes and face. However (yet again) my Nix stuff is not setup yet, and I have very little time to do it now. So if someone else can do it that'd be cool, otherwise I might do it in a few days (or weeks...)

File: test/test.hs (hunk 1 above)
Comment (roelvandijk) [inline test/test.hs]: We'll have to think about that path a bit. There will be more cases like this.

I don't know if the [standard Cabal solution](https://www.haskell.org/cabal/users-guide/developing-packages.html#accessing-data-files-from-package-code) works during development. That is, without installing the package.

**Comment by roelvandijk** [comment]
Very cool to have a Haar classifier. I'll look into the exampe when I have some spare time. Hopefully sometime next week.

I should note that the examples, at least those in the documentation, do not depend on a Nix setup. Just build the package with the test-suite enabled. `cabal configure --enable-tests` or whatever the `stack`  alternative is. Then you can run the image generating example code as `./dist/build/opencv-doc-images/opencv-doc-images`

File: test/test.hs (hunk 1 above)
Comment (bitonic) [inline test/test.hs]: Well, one thing we can do is to just include the `xml` file inside the package, if the license allows it.

**Comment by bitonic** [comment]
> I should note that the examples, at least those in the documentation, do not depend on a Nix setup. Just build the package with the test-suite enabled. `cabal configure --enable-tests` or whatever the `stack`  alternative is. Then you can run the image generating example code as `./dist/build/opencv-doc-images/opencv-doc-images`

Yeah, my problem is that I get
```
ghc: /usr/local/lib/libippicv.a: unknown symbol `__intel_cpu_features_init_x'
```
not sure why, but I hope nix will fix it.

File: test/test.hs (hunk 1 above)
Comment (roelvandijk) [inline test/test.hs]: I checked the license for the `xml` file. It is the BSD 3-Clause License with Intel Corporation as the copyright holder. Is see no problems with including it in this repository. Using it only in the test-suite and examples should free users from the responsibility of reproducing the license in binary redistributions.

Comment (bitonic) [inline test/test.hs]: I've imported them.

**Comment by basvandijk** [comment]
@bitonic regarding your `ghc: /usr/local/lib/libippicv.a: unknown symbol` error: you might want to try moving ippicv from the public to the private libs in the OpenCV pkgconfig file like I do [here](https://github.com/LumiGuide/haskell-opencv/blob/master/opencv-3.1.0.nix#L49).

**Comment by roelvandijk** [comment]
I merged this branch (and rebased a bit) in 6586467d0e39049deaff96980463828935425069. I also added an example as Francesco suggested.

Thanks Francesco!

**Comment by basvandijk** [comment]
Cool [example](http://lumiguide.github.io/haskell-opencv/doc/OpenCV-ImgProc-CascadeClassifier.html#v:cascadeClassifierDetectMultiScale)!

**Pull request closed without merging**